WordPress/Classes/Services/BlogService+Reminders.swift

```diff
@@ -1,15 +1,21 @@
 import Foundation
 
 extension BlogService {
-    @objc func unscheduleBloggingReminders(for blog: Blog) {
-        do {
-            let scheduler = try ReminderScheduleCoordinator()
-            scheduler.schedule(.none, for: blog, completion: { _ in })
-            // We're currently not propagating success / failure here, as it's
-            // it's only used when removing blogs or accounts, and there's
-            // no extra action we can take if it fails anyway.
-        } catch {
-            DDLogError("Could not instantiate the reminders scheduler: \(error.localizedDescription)")
+    @objc func unscheduleBloggingReminders(for blogID: NSManagedObjectID) {
+        let context = ContextManager.shared.mainContext
+        context.performAndWait {
+            do {
+                guard let blogInContext = try context.existingObject(with: blogID) as? Blog else {
+                    return
+                }
+                let scheduler = try ReminderScheduleCoordinator()
+                scheduler.schedule(.none, for: blogInContext, completion: { _ in })
```

Comment:
By dropping the completion handler we're effectively "firing and forgetting" this action – if we're calling it in a loop (as we do in `BlogService.m` below), how can we be certain that the schedule modification is complete prior to calling `context.deleteObject(blog)`? What happens if the blog object is deleted while `BloggingRemindersScheduler` is operating on it? We should be strictly controlling the order of operations here – by not doing so we're leaving room for bugs.

One alternative here could be to make a new `unschedule` method on `ReminderScheduleCoordinator` that takes a value type (and make this method use the same value type). That'd allow us to operate entirely within the context of the caller which would eliminate the possibility for Core Data issues (though we should still rework the IO to work properly async and propagate errors).

Reply:
> By dropping the completion handler we're effectively "firing and forgetting" this action
My knowledge of the Blogging Reminders code is a bit limited, but from some code tracing, I can see that the execution flow when unscheduling is synchronous. That's probably why the original author of this code decided to ignore the completion handler here. But I agree that this opens the chance of future bugs if this fact changes in the future.

> One alternative here could be to make a new unschedule method on ReminderScheduleCoordinator that takes a value type
Unfortunately, this won't be enough. The violation we're trying to fix (which you can find in #20848) actually happens in `WPAccount.lookup(withUUIDString:in:)`, where the main context is used to fetch the account while on a background context thread. The Blogging Reminders code, in general, utilizes the main context extensively, so using a value type instead of a `Blog` managed object won't cut it.

Reply:
```
Thread 38 Queue : NSManagedObjectContext 0x131a33400 (serial)
#0	0x00000001845323a0 in +[NSManagedObjectContext __Multithreading_Violation_AllThatIsLeftToUsIsHonor__] ()
#1	0x0000000184531f7c in _PFAssertSafeMultiThreadedAccess_impl ()
#2	0x0000000184539d54 in -[NSManagedObjectContext executeRequest:error:] ()
#3	0x000000018448f90c in NSManagedObjectContext.fetch<τ_0_0>(_:) ()
#4	0x000000010378ffa8 in static WPAccount.lookup(withUUIDString:in:) at /Users/garem/Documents/Projects/WordPress-iOS/WordPress/Classes/Models/WPAccount+Lookup.swift:52
#5	0x000000010378fd5c in static WPAccount.lookupDefaultWordPressComAccount(in:) at /Users/garem/Documents/Projects/WordPress-iOS/WordPress/Classes/Models/WPAccount+Lookup.swift:38
#6	0x000000010352f588 in BloggingPromptsService.init(contextManager:remote:blog:) at /Users/garem/Documents/Projects/WordPress-iOS/WordPress/Classes/Services/BloggingPromptsService.swift:185
```

I think the proper solution to this crash is fixing `BloggingPromptsService.init(contextManager:remote:blog:)`. It should not use `mainContext` to look up default account _without_ checking if it's safe to do that in its current thread. https://github.com/wordpress-mobile/WordPress-iOS/pull/20854 changed the initializer to execute the query correctly.

I had a similar thought as Jeremy's value type idea in https://github.com/wordpress-mobile/WordPress-iOS/pull/20854#pullrequestreview-1478347103. The whole blog reminder part doesn't seem to _need_ the `Blog` type (I'm not familiar with it either, so I could be wrong).

Reply:
👋🏼. Thanks for the input! As @crazytonyli said, through #20854 I've made some changes to `BloggingPromptsService.init(contextManager:remote:blog:)` that should address the default account violation.

However, I also agree that we need a larger revamp of the Blogging Reminder feature so that direct access to managed objects is limited and done from the correct queue. I've shared in https://github.com/wordpress-mobile/WordPress-iOS/pull/20854#issuecomment-1590913338 that this would require a larger refactor that involves a lot of changes. Since #20854 targets the frozen `release/22.6`, I've decided to keep its scope small enough to mitigate the identified violations so far, and do the larger refactor in the near future.

Reply:
@jkmassel @crazytonyli @dvdchr 
I took @dvdchr's #20854 for a spin to test the unscheduling flow, and I think I can build on top of it using the same patterns to fix the unscheduling violation 👍 

However, before I do so, can you take a look at #20853? It takes a more generic approach and should fix most of the violations. The idea is to pass down the background context to `BloggingPromptsService` and avoid the main context altogether. It may still need some refactoring to limit any potential misuse, but I'd love to hear your thoughts on the idea in general.

Reply:
> [...] pass down the background context to BloggingPromptsService and avoid the main context altogether

This is a pattern that we'd like to move away from. https://appsinfrap2.wordpress.com/2021/01/19/wordpress-ios-core-data-modernization-an-rfc/ has lots of details and background about the direction we'd like to head towards, which has been mostly implemented in the app now. Please see the "Don’t pass around `NSManagedObjectContext` objects on their own" section in particular.

Reply:
@crazytonyli That makes sense. And thanks for sharing this post! I'll close the two PRs and will work on a fix building on #20854 👍